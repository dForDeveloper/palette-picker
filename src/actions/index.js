export const setColors = (colors) => ({
  type: 'SET_COLORS',
  colors
});

export const setLockedColors = (lockedColors) => ({
  type: 'SET_LOCKED_COLORS',
  lockedColors
});

export const setProjects = (projects) => ({
  type: 'SET_PROJECTS',
  projects
});

export const toggleLoading = (bool) => ({
  type: 'TOGGLE_LOADING',
  bool
});

export const setError = (message) => ({
  type: 'SET_ERROR',
  message
});

export const setPalettes = (id, palettes) => ({
  type: 'SET_PALETTES',
  id,
  palettes
});

export const setModal = (isDisplayed, modalType = null, text = null) => ({
  type: 'SET_MODAL',
  isDisplayed,
  modalType,
  text
});