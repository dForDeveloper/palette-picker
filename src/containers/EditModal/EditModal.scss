.Modal--button-edit {
  @extend .Modal--button;

   &:disabled {
    background-color: #9facbd;
    cursor: not-allowed;
   }
}

.Modal--warning {
  text-align: center;
  font-size: .8rem;
  position: absolute;
  color: #ff0000;
  width: 256px;
}