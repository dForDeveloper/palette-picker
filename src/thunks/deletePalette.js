import { setError, toggleLoading, setModal, removePalette } from '../actions';
import { fetchData, createOptions } from '../utils/api';

export const deletePalette = (projectID, paletteID) => {
  return async (dispatch) => {
    dispatch(toggleLoading(true));
    try {
      const options = createOptions('DELETE');
      await fetchData(`/api/v1/palettes/${paletteID}`, options);
      dispatch(removePalette(projectID, paletteID));
      dispatch(setModal(false));
    } catch (error) {
      dispatch(setError('Error deleting palette: ' + error.message));
    }
    dispatch(toggleLoading(false));
  }
}