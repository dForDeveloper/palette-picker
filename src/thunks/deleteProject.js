import { setError, toggleLoading, removeProject, setModal } from '../actions';
import { fetchData, createOptions } from '../utils/api';

export const deleteProject = (id) => {
  return async (dispatch) => {
    dispatch(toggleLoading(true));
    try {
      const options = createOptions('DELETE');  
      await fetchData(`/api/v1/projects/${id}`, options);
      dispatch(removeProject(id));
      dispatch(setModal(false));
    } catch (error) {
      dispatch(setError('Error deleting project: ' + error.message));
    }
    dispatch(toggleLoading(false));
  }
}