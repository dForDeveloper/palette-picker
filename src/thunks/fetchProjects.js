import { setProjects, toggleLoading, setError } from '../actions';
import { fetchData } from '../utils/api';

export const fetchProjects = () => {
  return async (dispatch) => {
    dispatch(toggleLoading(true));
    try {
      const projects = await fetchData('/api/v1/projects');
      dispatch(setProjects(projects));
    } catch (error) {
      dispatch(setError('Error fetch projects: ' + error.message));
    }
    dispatch(toggleLoading(false))
  }
}