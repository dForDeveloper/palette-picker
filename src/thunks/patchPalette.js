import { setError, toggleLoading, setModal, updatePalette } from '../actions';
import { fetchData, createOptions } from '../utils/api';

export const patchPalette = (projectID, paletteID, name) => {
  return async (dispatch) => {
    dispatch(toggleLoading(true));
    try {
      const options = createOptions('PATCH', { name });
      await fetchData(`/api/v1/palettes/${paletteID}`, options);
      dispatch(updatePalette(projectID, paletteID, name));
      dispatch(setModal(false));
    } catch (error) {
      dispatch(setError('Error updating palette: ' + error.message));
    }
    dispatch(toggleLoading(false));
  }
}