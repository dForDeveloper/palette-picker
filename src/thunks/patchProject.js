import { setError, toggleLoading } from '../actions';
import { fetchData, createOptions } from '../utils/api';

export const patchProject = () => {
  return async (dispatch) => {
    dispatch(toggleLoading(true));
    try {
    
    } catch (error) {
      dispatch(setError('Error updating project: ' + error.message));
    }
    dispatch(toggleLoading(false));
  }
}