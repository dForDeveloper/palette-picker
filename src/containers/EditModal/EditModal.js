import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { patchProject } from '../../thunks/patchProject';
import { patchPalette } from '../../thunks/patchPalette';
import { deleteProject } from '../../thunks/deleteProject';
import { deletePalette } from '../../thunks/deletePalette';

export class EditModal extends Component {
  state = {
    editedName: '',
  };

  componentDidMount() {
    this.setState({ editedName: this.props.currentName });
  }

  handleChange = (event) => {
    this.setState({ editedName: event.target.value });
  };

  handleClick = () => {
    const { modalType, projectID, paletteID } = this.props;
    if (modalType === 'project') {
      this.props.deleteProject(projectID);
    } else if (modalType === 'palette') {
      this.props.deletePalette(projectID, paletteID);
    }
  }

  handleSubmit = (event) => {
    event.preventDefault();
    const { editedName } = this.state;
    const { modalType, projectID, paletteID } = this.props;
    if (modalType === 'project') {
      this.props.patchProject(projectID, editedName);
    } else if (modalType === 'palette') {
      this.props.patchPalette(projectID, paletteID, editedName);
    }
  }

  render() {
    const { projects, currentName, modalType } = this.props;
    const saveIsDisabled = modalType === 'project' && projects
      .filter(project => {
        return project.name !== currentName;
      })
      .map(project => project.name)
      .includes(this.state.editedName);
    return (
      <form className="Modal" onSubmit={this.handleSubmit}>
        <h4 className="Modal--h4">{`edit ${this.props.modalType}`}</h4>
        <input
          className="Modal--input"
          placeholder={`${this.props.modalType} name`}
          value={this.state.editedName}
          onChange={this.handleChange}
          autoFocus={true}
          required 
        />
        {saveIsDisabled && 
          <p className="Modal--warning">projects cannot have the same name</p>}
        <input
          type="submit"
          value="save"
          className="Modal--button-edit"
          disabled={saveIsDisabled}
        />
        <button className="Modal--button" onClick={this.handleClick}>
          delete
        </button>
      </form>
    );
  }
}

export const mapStateToProps = (state) => ({
  projects: state.projects
});

export const mapDispatchToProps = (dispatch) => ({
  patchProject: (projectID, editedName) => {
    dispatch(patchProject(projectID, editedName))
  },
  patchPalette: (projectID, paletteID, editedName) => {
    dispatch(patchPalette(projectID, paletteID, editedName))
  },
  deleteProject: (id) => dispatch(deleteProject(id)),
  deletePalette: (projectID, paletteID) => {
    dispatch(deletePalette(projectID, paletteID))
  }
});

export default connect(mapStateToProps, mapDispatchToProps)(EditModal);

EditModal.propTypes = {
  isDisplayed: PropTypes.bool,
  modalType: PropTypes.string,
  currentName: PropTypes.string,
  projectID: PropTypes.number,
  paletteID: PropTypes.number,
  projects: PropTypes.array,
  patchProject: PropTypes.func,
  patchPalette: PropTypes.func,
  deleteProject: PropTypes.func,
  deletePalette: PropTypes.func,
};