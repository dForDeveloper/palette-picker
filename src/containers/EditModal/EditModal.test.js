import React from 'react';
import { shallow } from 'enzyme';
import { setColors, setModal } from '../../actions';
import { patchProject } from '../../thunks/patchProject';
import { patchPalette } from '../../thunks/patchPalette';
import { deleteProject } from '../../thunks/deleteProject';
import { deletePalette } from '../../thunks/deletePalette';
import { EditModal, mapStateToProps, mapDispatchToProps } from './EditModal';

jest.mock('../../thunks/patchProject');
jest.mock('../../thunks/patchPalette');
jest.mock('../../thunks/deleteProject');
jest.mock('../../thunks/deletePalette');

describe('EditModal', () => {
  let wrapper;
  const mockProjectProps = {
    isDisplayed: true,
    modalType: 'project',
    currentName: 'my project',
    projectID: 1,
    paletteID: null,
    projects: [{ name: 'my project' }, { name: 'other project'}],
    patchProject: jest.fn(),
    patchPalette: jest.fn(),
    deleteProject: jest.fn(),
    deletePalette: jest.fn()
  };
  const mockPaletteProps = {
    ...mockProjectProps,
    modalType: 'palette',
    currentName: 'my palette',
    projectID: 1,
    paletteID: 1
  };

  beforeEach(() => {
    wrapper = shallow(<EditModal {...mockProjectProps} />);
  });

  describe('render', () => {
    it('should match the snapshot when modalType is project', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should match the snapshot when modalType is palette', () => {
      wrapper = shallow(<EditModal {...mockPaletteProps} />);
      expect(wrapper).toMatchSnapshot();
    });

    it('should match the snapshot is saveIsDisabled', () => {
      wrapper.setState({ editedName: 'other project' });
      expect(wrapper).toMatchSnapshot();
    });

    it('should call handleSubmit when the form is submitted', () => {
      jest.spyOn(wrapper.instance(), 'handleSubmit');
      wrapper.instance().forceUpdate();
      const mockEvent = { preventDefault: jest.fn() };
      wrapper.find('.Modal').simulate('submit', mockEvent);
      expect(wrapper.instance().handleSubmit).toHaveBeenCalled();
    });

    it('should call handleChange when the input changes', () => {
      jest.spyOn(wrapper.instance(), 'handleChange');
      wrapper.instance().forceUpdate();
      const mockEvent = { target: { value: 'a'} };
      wrapper.find('.Modal--input').simulate('change', mockEvent);
      expect(wrapper.instance().handleChange).toHaveBeenCalled();
    });

    it('should call handleClick the delete button is clicked', () => {
      jest.spyOn(wrapper.instance(), 'handleClick');
      wrapper.instance().forceUpdate();
      wrapper.find('.Modal--button').simulate('click');
      expect(wrapper.instance().handleClick).toHaveBeenCalled();
    });
  });

  describe('componentDidMount', () => {
    it('should set state with the currentName', () => {
      wrapper.instance().componentDidMount();
      expect(wrapper.state('editedName')).toEqual('my project');
    });
  });

  describe('handleChange', () => {
    it('should set state with event.target.value', () => {
      const mockEvent = { target: { value: 'qwerty' } };
      wrapper.instance().handleChange(mockEvent);
      expect(wrapper.state('editedName')).toEqual('qwerty');
    });
  });

  describe('handleClick', () => {
    it('should call deleteProject if modalType is project', () => {
      wrapper.instance().handleClick();
      expect(mockProjectProps.deleteProject).toHaveBeenCalled();
    });

    it('should call deletePalette if modalType is palette', () => {
      wrapper = shallow(<EditModal {...mockPaletteProps} />);
      wrapper.instance().handleClick();
      expect(mockPaletteProps.deletePalette).toHaveBeenCalled();
    });
  });

  describe('handleSubmit', () => {
    it('should call patchProject if modalType is project', () => {
      wrapper.instance().handleSubmit({ preventDefault: jest.fn() });
      expect(mockProjectProps.patchProject).toHaveBeenCalled();
    });

    it('should call patchPalette if modalType is palette', () => {
      wrapper = shallow(<EditModal {...mockPaletteProps} />);
      wrapper.instance().handleSubmit({ preventDefault: jest.fn() });
      expect(mockPaletteProps.patchPalette).toHaveBeenCalled();
    });
  });
});

describe('mapStateToProps', () => {
  it('should return a props object with the correct properties', () => {
    const mockState = {
      projects: [{ id: 1, name: 'my project' }],
      extraProperty: true
    };
    const expected = { projects: [{ id: 1, name: 'my project' }] };
    const result = mapStateToProps(mockState);
    expect(result).toEqual(expected);
  });
});

describe('mapDispatchToProps', () => {
  const mockDispatch = jest.fn();
  const mappedProps = mapDispatchToProps(mockDispatch);

  it('should call patchProject', () => {
    const expected = patchProject(1, 'edited project name');
    mappedProps.patchProject(1, 'edited project name');
    expect(mockDispatch).toHaveBeenCalledWith(expected);
  });

  it('should call patchPalette', () => {
    const expected = patchPalette(1, 2, 'edited palette name');
    mappedProps.patchPalette(1, 2, 'edited palette name');
    expect(mockDispatch).toHaveBeenCalledWith(expected);
  });

  it('should call deleteProject', () => {
    const expected = deleteProject(1);
    mappedProps.deleteProject(1);
    expect(mockDispatch).toHaveBeenCalledWith(expected);
  });

  it('should call deletePalette', () => {
    const expected = deletePalette(1, 2);
    mappedProps.deletePalette(1, 2);
    expect(mockDispatch).toHaveBeenCalledWith(expected);
  });
});