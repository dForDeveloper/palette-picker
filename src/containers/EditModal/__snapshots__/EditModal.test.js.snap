// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditModal render should match the snapshot is saveIsDisabled 1`] = `
<form
  className="Modal"
  onSubmit={[Function]}
>
  <h4
    className="Modal--h4"
  >
    edit project
  </h4>
  <input
    autoFocus={true}
    className="Modal--input"
    onChange={[Function]}
    placeholder="project name"
    required={true}
    value="other project"
  />
  <p
    className="Modal--warning"
  >
    projects cannot have the same name
  </p>
  <input
    className="Modal--button-edit"
    disabled={true}
    type="submit"
    value="save"
  />
  <button
    className="Modal--button"
    onClick={[Function]}
  >
    delete
  </button>
</form>
`;

exports[`EditModal render should match the snapshot when modalType is palette 1`] = `
<form
  className="Modal"
  onSubmit={[Function]}
>
  <h4
    className="Modal--h4"
  >
    edit palette
  </h4>
  <input
    autoFocus={true}
    className="Modal--input"
    onChange={[Function]}
    placeholder="palette name"
    required={true}
    value="my palette"
  />
  <input
    className="Modal--button-edit"
    disabled={false}
    type="submit"
    value="save"
  />
  <button
    className="Modal--button"
    onClick={[Function]}
  >
    delete
  </button>
</form>
`;

exports[`EditModal render should match the snapshot when modalType is project 1`] = `
<form
  className="Modal"
  onSubmit={[Function]}
>
  <h4
    className="Modal--h4"
  >
    edit project
  </h4>
  <input
    autoFocus={true}
    className="Modal--input"
    onChange={[Function]}
    placeholder="project name"
    required={true}
    value="my project"
  />
  <input
    className="Modal--button-edit"
    disabled={false}
    type="submit"
    value="save"
  />
  <button
    className="Modal--button"
    onClick={[Function]}
  >
    delete
  </button>
</form>
`;
